### PR TITLE
Add support for parameter optionality

### PR DESCRIPTION
**Pull request by captainsafia**
🥚 🍳

### CODE REVIEW

File: src/uController.SourceGenerator/Reflection/RoslynType.cs [hunk 1]
```diff
@@ -40,13 +40,13 @@ public RoslynType(ITypeSymbol typeSymbol, MetadataLoadContext metadataLoadContex
 
         public override string Name => ArrayTypeSymbol is { } ar ? ar.ElementType.MetadataName + "[]" : _typeSymbol.MetadataName;
 
-        public override bool IsGenericType => NamedTypeSymbol?.IsGenericType ?? false;
+        public override bool IsGenericType => (NamedTypeSymbol?.IsGenericType ?? false) || _typeSymbol.NullableAnnotation == NullableAnnotation.Annotated;
```
Comment (davidfowl) [inline src/uController.SourceGenerator/Reflection/RoslynType.cs]: I don't understand this.

File: src/uController.SourceGenerator/Reflection/RoslynType.cs [hunk 2]
```diff
@@ -77,6 +80,10 @@ public override Type[] GetGenericArguments()
 
         public override Type GetGenericTypeDefinition()
         {
+            if (NamedTypeSymbol.NullableAnnotation == NullableAnnotation.Annotated)
+            {
+                return typeof(Nullable<>);
+            }
```
Comment (davidfowl) [inline src/uController.SourceGenerator/Reflection/RoslynType.cs]: This doesn't seem right. `Nullable<T>` is an actual type while nullable annotations are compile time magic.

File: src/uController.SourceGenerator/uControllerGenerator.cs [hunk 3]
```diff
@@ -152,6 +152,7 @@ IOperation ResolveDeclarationOperation(ISymbol symbol)
                     {
                         Method = methodModel,
                         ParameterSymbol = parameter.GetParameterSymbol(),
+                        ParameterInfo = new RoslynParameterInfo(parameter.GetParameterSymbol(), metadataLoadContext),
```
Comment (davidfowl) [inline src/uController.SourceGenerator/uControllerGenerator.cs]:
```suggestion
                        ParameterInfo = parameter,
```

File: src/uController.SourceGenerator/Reflection/RoslynType.cs (hunk 2 above)
Comment (captainsafia) [reply src/uController.SourceGenerator/Reflection/RoslynType.cs]: 💡 Yeah, an example of where I got a little caught up trying to follow the reflection-style patterns in this API. I had this in so I could derive the unannotated state of nullable types for when they're passed into `typeof` but I realize that's not necessary.

File: src/uController.SourceGenerator/MinimalCodeGenerator.cs [hunk 4]
```diff
@@ -330,39 +359,53 @@ private void EmitParameter(ref bool hasAwait, ref bool hasFromBody, ref bool has
...
                     parameter.Unresovled = true;
                 }
             }
             else if (parameter.FromServices)
             {
-                WriteLine($"var {parameterName} = httpContext.RequestServices.GetRequiredService<{parameter.ParameterType}>();");
+                if (isOptional)
+                {
+                    WriteLine(
+                              $"var {parameterName} = httpContext.RequestServices.GetService<{parameter.ParameterType}>();");
```
Comment (davidfowl) [inline src/uController.SourceGenerator/MinimalCodeGenerator.cs]:
```suggestion
                    WriteLine($"var {parameterName} = httpContext.RequestServices.GetService<{parameter.ParameterType}>();");
```

File: src/uController.SourceGenerator/MinimalCodeGenerator.cs [hunk 5]
```diff
@@ -330,39 +359,53 @@ private void EmitParameter(ref bool hasAwait, ref bool hasFromBody, ref bool has
...
+                }
+                else
+                {
+                    WriteLine(
+                              $"var {parameterName} = httpContext.RequestServices.GetRequiredService<{parameter.ParameterType}>();");
```
Comment (davidfowl) [inline src/uController.SourceGenerator/MinimalCodeGenerator.cs]:
```suggestion
                    WriteLine($"var {parameterName} = httpContext.RequestServices.GetRequiredService<{parameter.ParameterType}>();");
```

File: src/uController.SourceGenerator/MinimalCodeGenerator.cs [hunk 6]
```diff
@@ -372,6 +415,15 @@ private void EmitParameter(ref bool hasAwait, ref bool hasFromBody, ref bool has
             else if (parameter.FromBody)
             {
                 BodyParameters.Add(parameter);
+                var fromBodyAttribute =
+                    parameter.ParameterInfo.GetCustomAttributeData(_wellKnownTypes.FromBodyMetadataType);
```
Comment (davidfowl) [inline src/uController.SourceGenerator/MinimalCodeGenerator.cs]: Maybe store it on the parameter object?

File: src/uController.SourceGenerator/MinimalCodeGenerator.cs [hunk 7]
```diff
@@ -383,8 +435,8 @@ private void EmitParameter(ref bool hasAwait, ref bool hasFromBody, ref bool has
                 }
                 else
                 {
-
-                    WriteLine($"var {parameterName} = await ResolveBody<{parameter.ParameterType}>(httpContext);");
+                    WriteLine(
```
Comment (davidfowl) [inline src/uController.SourceGenerator/MinimalCodeGenerator.cs:438]: Why this formattig change

File: src/uController.SourceGenerator/MinimalCodeGenerator.cs [hunk 8]
```diff
@@ -383,8 +435,8 @@ private void EmitParameter(ref bool hasAwait, ref bool hasFromBody, ref bool has
...
+                              $"var {parameterName} = await ResolveBody<{parameter.ParameterType}>(httpContext, {(isOptional ? "true" : "false")});");
```
Comment (davidfowl) [inline src/uController.SourceGenerator/MinimalCodeGenerator.cs]: Why this formatting change?

```suggestion
                    WriteLine($"var {parameterName} = await ResolveBody<{parameter.ParameterType}>(httpContext, {(isOptional ? "true" : "false")});");
```

**Comment by davidfowl** [comment]
This PR has way too many formatting changes. Makes it really hard to review. The core of this change seems small. Can you revert those changes and make them in a follow up (well some of them).

File: src/uController.SourceGenerator/MinimalCodeGenerator.cs [hunk 9]
```diff
@@ -453,49 +524,78 @@ private bool HasTryParseMethod(Type type, out MethodInfo mi)
...
-                        WriteLine($@"var {sourceName} = {getter}?.ToString() ?? ""{defaultValue}"";");
+                        WriteLine($@"var {sourceName} = getter{sourceName}Value?.ToString() ?? ""{defaultValue}"";");
                     }
                     else
                     {
-                        WriteLine($@"var {sourceName}Str = {getter}.ToString();");
-                        WriteLine($@"var {sourceName} = string.IsNullOrEmpty({sourceName}Str) ? ""{defaultValue}"" : {sourceName}Str;");
+                        WriteLine($@"var {sourceName}Str = getter{sourceName}Value.ToString();");
+                        WriteLine(
+                                  $@"var {sourceName} = string.IsNullOrEmpty({sourceName}Str) ? ""{defaultValue}"" : {sourceName}Str;");
```
Comment (davidfowl) [inline src/uController.SourceGenerator/MinimalCodeGenerator.cs]: missed one.

```suggestion
                        WriteLine($@"var {sourceName} = string.IsNullOrEmpty({sourceName}Str) ? ""{defaultValue}"" : {sourceName}Str;");
```

File: src/uController.SourceGenerator/MinimalCodeGenerator.cs [hunk 10]
```diff
@@ -570,7 +673,7 @@ private void GenerateTryParse(MethodInfo tryParseMethod, string sourceName, stri
                 }
                 else
                 {
-                    WriteLine($"if ({sourceName} == null)");
+                    WriteLine($"if (string.IsNullOrEmpty({sourceName}))");
```
Comment (davidfowl) [inline src/uController.SourceGenerator/MinimalCodeGenerator.cs:673]: Is this right?

Comment (captainsafia) [reply src/uController.SourceGenerator/MinimalCodeGenerator.cs:673]: I believe so. Reason for this is the discrepancy between what you get out of `ResolveByQueryOrRoute` when no parameter is provided. For the query scenario, you'll get an empty `StringValue`. For the route scenario, you'll get `null` so the previous scenario was only handling the route scenario successfully.

**Review by davidfowl — APPROVED** [review]

**Pull request merged**